helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java

```diff
@@ -201,6 +217,58 @@ protected void handleEvent(NotificationContext event) {
         logger.warn("Exception in callback processing thread. Skipping callback", e);
       }
     }
+
+    public boolean queueIsEmpty() {
+      return _eventQueue.isEmpty();
+    }
+  }
+
+  /**
+   * This class represents each periodic refresh task
+   * If the deadline passed, enqueue an event to do a refresh
+   * If the deadline has not passed (some events happened in between so don't need to do a refresh),
+   * wait until reaching the interval and check again
+   */
+  class TriggerTask implements Runnable {
+    private final long _triggerInterval;
+
+    public TriggerTask(long triggerInterval) {
+      this._triggerInterval = triggerInterval;
+    }
+
+    @Override
+    public void run() {
+      logger.info("Starting trigger task.");
+      while (!Thread.currentThread().isInterrupted()) {
```

Comment:
As we discussed this may cause the runnable to occupy the thread forever, let's try the schedule executor back.
Unfortunately, the simpler design fails the tests. I hope the older design will work.